### DISCUSSION TITLE
Small note about `:has()` performance

### DISCUSSION THREAD

**Discussion by ghost** [uBlockOrigin/uAssets#23949] [General]
https://github.com/refined-github/refined-github/issues/7331#issuecomment-2120464245

> > Is there an issue with the selector? Can that be improved too?
> 
> I wasn't entirely sure myself. After chatting with Mozillians about it, they said that anchoring `:has()` to an element (or to all elements) slows down all DOM mutations inside that subtree (with more performance impact the more complex the selector inside `:has()` is). So stuff like `body:has()` or `:root:has()` or `*:has()` are recipes for bad performance.
> 
> For example, the [`quick-mention`](https://github.com/refined-github/refined-github/blob/main/source/features/quick-mention.tsx) feature [uses](https://github.com/refined-github/refined-github/blob/23386186651562c6591790128d3bcf0c73a60185/source/features/quick-mention.tsx#L98) `body:has()`. It would likely be better to use a more specific anchor than `body` here. It's perhaps even better to avoid using `has()` if a simpler selector would suffice.

***

And https://github.com/refined-github/refined-github/pull/7378#issue-2326995802

**Comment by gorhill** [comment]
Related bugzilla issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1896380

**Comment by stephenhawk8054** [comment]
So, do filters like `##div:has()` have the same issue as the above CSS selectors? We are having ~30 filters like that in uAssets and EL.

**Reply by Yuki2718** [reply]
If I understand correctly, the problem is counts on subtree (and thus `body:has()` is bad) but in general won't be good.
